### PR TITLE
Enabled deploy_groups

### PR DESCRIPTION
**Pull request by mjksmith**
Placing multiple instances in a deploy group makes it so they are all deployed with the same docker container and are deployed at the same time (as specified in deploy.yaml). However, they can be controlled independently by flagging deploy_state and force_bounce in git refs.

### CODE REVIEW

**Comment by solarkennedy** [comment]
Sorry, lots of widespread changes. Can you rebase?

File: paasta_tools/cli/cmds/mark_for_deployment.py [hunk 1]
```diff
@@ -50,8 +50,8 @@ def add_subparser(subparsers):
         required=True,
     )
     list_parser.add_argument(
-        '-l', '--clusterinstance',
-        help='Mark the service ready for deployment in this clusterinstance (e.g. '
+        '-l', '--deploy_group',
```
Comment (solarkennedy) [inline paasta_tools/cli/cmds/mark_for_deployment.py]: You are going to have to make this backwards compatible. All of our existing jenkins jobs are doing --clusterinstance right now

File: paasta_tools/cli/cmds/mark_for_deployment.py [hunk 2]
```diff
@@ -48,8 +48,8 @@ def add_subparser(subparsers):
         required=True,
     )
     list_parser.add_argument(
-        '-l', '--clusterinstance',
-        help='Mark the service ready for deployment in this clusterinstance (e.g. '
+        '-l', '--deploy_group', '--clusterinstance',
+        help='Mark the service ready for deployment in this deploy group (e.g. '
              'cluster1.canary, cluster2.main)',
```
Comment (EvanKrall) [inline paasta_tools/cli/cmds/mark_for_deployment.py]: Make a note here that `--clusterinstance` is deprecated in favor of `--deploy_group`; I'm in favor of backwards compatibility here, but we should clean up the old option name at some point.

File: paasta_tools/utils.py [hunk 3]
```diff
@@ -971,8 +984,12 @@ def get_branch_dict(self, service, branch):
         return self.get(full_branch, {})
 
 
+def get_paasta_branch_from_identifier(identifier):
```
Comment (EvanKrall) [inline paasta_tools/utils.py]: I feel like we should just call it `deploy_group` instead of `identifier`, and thus `get_paasta_branch_from_deploy_group`.

File: paasta_tools/cli/cmds/check.py [hunk 4]
```diff
@@ -209,10 +211,16 @@ def get_chronos_steps(service, soa_dir):
     with what deploy.yaml has for linting."""
     steps = []
     for cluster in list_clusters(service, soa_dir):
-        for instance in get_service_instance_list(service=service, cluster=cluster,
-                                                  instance_type='chronos',
-                                                  soa_dir=soa_dir):
-            steps.append("%s.%s" % (cluster, instance[1]))
+        for _, instance in get_service_instance_list(service=service, cluster=cluster,
```
Comment (Rob-Johnson) [inline paasta_tools/cli/cmds/check.py]: I don't know if pre-commit will pick this up, but can you align these?

File: paasta_tools/cli/cmds/rollback.py [hunk 5]
```diff
@@ -65,57 +60,53 @@ def add_subparser(subparsers):
...
-    cluster, instance and sha. These arguments will be verified and passed onto
+    deploy groups and sha. These arguments will be verified and passed onto
     mark_for_deployment.
     """
     service = figure_out_service_name(args)
-    cluster = args.cluster
     git_url = get_git_url(service)
     commit = args.commit
-    given_instances = args.instances.split(",")
+    given_deploy_groups = filter(bool, args.deploy_groups.split(","))
```
Comment (Rob-Johnson) [inline paasta_tools/cli/cmds/rollback.py]: This is...confusing. What are you trying to achieve here?

File: paasta_tools/cli/cmds/rollback.py [hunk 6]
```diff
@@ -65,57 +60,53 @@ def add_subparser(subparsers):
...
-            returncode = mark_for_deployment(
-                git_url=git_url,
-                cluster=cluster,
-                instance=instance,
-                service=service,
-                commit=commit,
-            )
-    else:
-        print PaastaColors.red("ERROR: The service %s is not deployed into cluster %s.\n" % (service, cluster))
+    if len(deploy_groups) is 0:
```
Comment (Rob-Johnson) [inline paasta_tools/cli/cmds/rollback.py]: use ``==`` rather than ``is`` here

**Comment by Rob-Johnson** [comment]
First pass, I'm gonna read this again.

File: paasta_tools/cli/cmds/rollback.py (hunk 5 above)
Comment (mjksmith) [inline paasta_tools/cli/cmds/rollback.py]: The `instances` parameter to paasta rollback is supposed to be optional; if it's not included rollback should rollback every instance.

This code removes empty strings from `given_instances`. Before this if the instances argument wasn't passed to rollback it would default to `''`, so given_instances would be `['']`. This causes rollback not to work properly as the rest of the code checks to see if `len(given_instances) == 0`.

I chose to make this change rather than changing the logic for determining if `given_instances` is empty because this also deals with some typos (correcting multiple commas like `'instance1,,instance2'`) and also removing trailing commas (before `'instance_name,'` would resolve to `['instance_name', '']`).

TL;DR this feature was always broken and filtering with the `bool` function removes empty strings from the result of `split`.

Comment (Rob-Johnson) [inline paasta_tools/cli/cmds/rollback.py]: I see. This still fails if you do ``instance1, ,instance2`` though:

```python
> bool(' ')
True
> filter(bool, "instance1, ,instance2".split(","))
['instance1', ' ', 'instance2']
```
How about:

```python
# filter the empty strings
[deploy_group for deploy_group in args.deploy_groups.split(",") if not deploy_group]
```
?

Yeah, it's probably slower, but the cost of that vs the non obvious behaviour of this code is worth it imo. What do you think?

Comment (mjksmith) [inline paasta_tools/cli/cmds/rollback.py]: I'm not actually sure how argparse will handle spaces in an argument.

I still like your list comprehension more though, I'll make that change.
